### ISSUE TITLE
[bug] Game crashes on custom campaign

### ISSUE THREAD

**Issue by Cleeb** [wesnoth/wesnoth#2116]
I am running Wesnoth 1.12.6 from the Vincent Cheng ppa on Ubuntu 16.04.  I am working on my own campaign and on the fifth scenario it will crash at what seems to be the end of side two's turn and the start of side three's.  This bug seems extremely inconsistent and I will sometimes be able to play through the scenario with nothing out of the ordinary happening, at other times the entire game will crash.  I have attached the campaign, game log, and two savegame files that demonstrate the bug that is occurring.  In the savegame "Rew-The Drakes-Turn 6" if you end the turn the game will crash after side two's turn as long as you don't move the unit named Treyce.  In the savegame "Rew-The Drakes-Turn 6test" I have moved the unit named Treyce already and the game will not crash.   

[Revolutioncrash.zip](https://github.com/wesnoth/wesnoth/files/1404994/Revolutioncrash.zip)

**Comment by gfgtdf** [comment]
here is the assertion message in his logs (in that zifile)
`wesnoth: /build/wesnoth-1.12-BTrJSk/wesnoth-1.12-1.12.6/src/ai/testing/../../unit_map.hpp:176: unit_map::iterator_base<iter_types>::value_type* unit_map::iterator_base<iter_types>::operator->() const [with iter_types = unit_map::const_iter_types; unit_map::iterator_base<iter_types>::pointer = const unit*; unit_map::iterator_base<iter_types>::value_type = const unit]: Assertion `valid()' failed.`

**Comment by Cleeb** [comment]
Sorry, there was a mistake in the files I uploaded.  In these files I have the campaign start on the scenario that crashes, in the old one it started on a scenario that I was using to help fix an older campaign. [Revolutioncrash.zip](https://github.com/wesnoth/wesnoth/files/1452154/Revolutioncrash.zip)

**Comment by Cleeb** [comment]
I figured it out!  I think.  I decided to simply remake the scenario in hopes that it would fix the problem.  It did but I had no idea why. I made things pretty much the same but with one real difference; I had given all of the leaders their own side, so there were no longer multiple leaders on side 2.  I didn't think twice of it until the same problem occurred on a later scenario.  In this scenario I had two units that could recruit on the same side and I had the same problem.  I moved the other leader to a new side and the problem was fixed.  I have no idea why having more than one leader is doing this but I am almost certain that is what is causing the crash.
